input/script_utils.go

```diff
@@ -443,6 +479,255 @@ func SenderHtlcSpendTimeout(receiverSig Signature,
+		revokeKey, tapScriptRoot[:],
+	)
+
+	return &HtlcScriptTree{
+		TaprootKey:     htlcKey,
+		SuccessTapLeaf: successTapLeaf,
+		TimeoutTapLeaf: timeoutTapLeaf,
+		TapscriptTree:  tapscriptTree,
+	}, nil
+}
+
+// SenderHTLCScriptTaproot constructs the taproot witness program (schnorr key)
+// for an outgoing HTLC on the sender's version of the commitment transaction.
+// This method returns the top level tweaked public key that commits to both
+// the script paths.
+//
+// The returned key commits to a tapscript tree with two possible paths:
+//
+//   - Timeout path:
+//     <local_key> OP_CHECKSIGVERIFY
+//     <remote_key> OP_CHECKSIG
+//
+//   - Success path:
+//     OP_SIZE 32 OP_EQUALVERIFY
+//     OP_HASH160 <RIPEMD160(payment_hash)> OP_EQUALVERIFY
+//     <remote_htlcpubkey> OP_CHECKSIG
+//     OP_CHECKSEQUENCEVERIFY
+//
+// The timeout path can be spent with a witness of (sender timeout):
+//
+//	<receiver sig> <local sig> <timeout_script> <control_block>
+//
+// The success path can be spent with a valid control block, and a witness of
+// (receiver redeem):
+//
+//	<receiver sig> <preimage> <success_script> <control_block>
+//
+// The top level keyspend key is the revocation key, which allows a defender to
+// unilaterally spend the created output.
+func SenderHTLCScriptTaproot(senderHtlcKey, receiverHtlcKey,
+	revokeKey *btcec.PublicKey, payHash []byte) (*HtlcScriptTree, error) {
+
+	// Given all the necessary parameters, we'll return the HTLC script
+	// tree that includes the top level output script, as well as the two
+	// tap leaf paths.
+	return senderHtlcTapScriptTree(
+		senderHtlcKey, receiverHtlcKey, revokeKey, payHash,
+	)
+}
+
+// SenderHTLCScriptTaprootRedeem creates a valid witness needed to redeem a
+// sender taproot HTLC with the pre-image. The returned witness is valid and
+// includes the control block required to spend the output.
+func SenderHTLCScriptTaprootRedeem(signer Signer, signDesc *SignDescriptor,
+	sweepTx *wire.MsgTx, preimage []byte, revokeKey *btcec.PublicKey,
+	tapscriptTree *txscript.IndexedTapScriptTree) (wire.TxWitness, error) {
+
+	sweepSig, err := signer.SignOutputRaw(sweepTx, signDesc)
+	if err != nil {
+		return nil, err
+	}
+
+	// In addition to the signature and the witness/leaf script, we also
+	// need to make a control block proof using the tapscript tree.
+	successTapLeafHash := txscript.NewBaseTapLeaf(
+		signDesc.WitnessScript,
+	).TapHash()
+	successIdx := tapscriptTree.LeafProofIndex[successTapLeafHash]
+	successMerkleProof := tapscriptTree.LeafMerkleProofs[successIdx]
+	successControlBlock := successMerkleProof.ToControlBlock(revokeKey)
+
+	// The final witness stack is:
+	//  <receiver sig> <preimage> <timeout_script> <control_block>
+	witnessStack := make(wire.TxWitness, 4)
+	witnessStack[0] = append(sweepSig.Serialize(), byte(signDesc.HashType))
+	witnessStack[1] = preimage
+	witnessStack[2] = signDesc.WitnessScript
+	witnessStack[4], err = successControlBlock.ToBytes()
```

Comment:
Should be index 3?

Reply:
Yep, tests forthcoming!

Reply:
this is index 4, so unresolving

Reply:
@Crypt-iQ - all these nits are fixed in the last commit of the PR (I think the idea is that they will be squashed in once fully reviewed)